### ISSUE TITLE
Adding the webapp's badge to the manifest

### ISSUE THREAD

**Issue by beverloo** [w3c/manifest#480]
I'd like to propose adding a `badge` property to the Manifest.

Authors would use this property to indicate a small icon that should be used to represent their webapp in space constrained locations, for example the status bar of a mobile device. Such badges are consistently displayed differently from regular icons, so using the smallest `icons` value (or a favicon for that matter) does not seem appropriate.

Today, both Chrome and Firefox show their respective logos in the status bar when showing a notification. This portrays no information to the user other regarding the source of the notification.

Web Notifications have a property called `badge` [(spec)](https://notifications.spec.whatwg.org/#badge-url) in which the developer can provide their own icon, but this is still significantly constrained by platform limitations (only ~13% of Android users have the necessary platform support).

In addition, there are various situations in which the browser shows notifications on behalf of the webapp.

Allowing authors to specify their webapp's badge in the manifest enables a more consistent experience for such browser-level notifications. Knowing which icon to badge notifications with at installation time makes it possible to work around these platform limitations, too.

Thoughts?

**Comment by benfredwells** [comment]
Some platforms (e.g. android) show a monochrome badge. Any ideas of how we'd allow sites to specify a monochrome badge and a color one?

**Comment by beverloo** [comment]
Different platforms take completely different approaches in displaying badges. You mention Android, it's not only monochrome, but it also limits display to non-transparent pixels. (Discarding partial transparency.) The actual colour may also display on, for example, theming.

Notifications has the following note that, incidentally, should be updated to mention badges as well.

> Some platforms may modify an action icon resource to better match the platform’s visual style before displaying it to the user, for example by rounding the corners or painting it in a specific color. It is recommended to use an icon that handles such cases gracefully and does not lose important information through, e.g., loss of color or clipped corners.

**Comment by marcoscaceres** [comment]
We had related discussions to this previously on the WHATWG list.
https://www.mail-archive.com/whatwg%40lists.whatwg.org/msg38109.html

And:
https://www.mail-archive.com/www-archive@w3.org/msg05623.html

tl;dr: we need "badged-tasks API" for this (yes, with a less terrible name), where those tasks are reflected as badge numbers or similar. I'm traveling this week, so can only write a short reply. Would be great to discuss further.

**Comment by kenchris** [comment]
@marcoscaceres I don't think this is about showing a number badge, like unread emails.

@beverloo can you show some screenshots of where you intend this to be used? It is like an overlay right, like a gmail logo on top of a profile pic etc, right?

**Comment by beverloo** [comment]
Correct, this talks about _badges_ as in _status bar icons_, following Notification API terminology.

I took two screenshots on Android, but it's effectively the same on all mobile platforms. Desktop platforms may choose to badge their toasts as well for additional attribution.

[Badges in the status bar](https://static.peter.sh/stuff/android-status-bar.png), and [badges on the notifications](https://static.peter.sh/stuff/android-notifications.png).

Again, this issue is about being able to provide an icon that replaces the UA's icon in these scenarios with something appropriate for the webapp, in situations where the developer currently can't.

**Comment by skddc** [comment]
Seems to me like this is very similar to the monochrome SVG icon that e.g. desktop Safari uses for their (fairly) new Pinned Sites (in the tab bar). So would it perhaps be an option to solve it via additional attributes or attribute values for `icons` items?

**Comment by marcoscaceres** [comment]
Ah, apologies for misunderstanding.

**Comment by kenchris** [comment]
So these are basically black/white (and inverted) images. Would be nice to specify these in a cross platform way.

**Comment by marcoscaceres** [comment]
So there is a need for these at least on Android. On iOS tho, the the app icons are used. 

![screenshot 2016-08-04 17 19 40](https://cloud.githubusercontent.com/assets/870154/17393550/b055c9e0-5a67-11e6-843a-d0e7c2e147eb.png)

However, couldn't this just be added when the notification is created instead of having a generic icon? Or is there no option for that on Android (i.e., you have to specify the application's icon for notification up-front)?

**Comment by beverloo** [comment]
Browsers currently use their own icons on Android, but that's definitely not sufficient for branding. That's why the per-notification option was introduced. This is insufficient in two cases:

1. Notifications shown on behalf of the website, but not through the Notification API. (E.g. update notifications, media notifications, Chrome's forced push notification.)
2. Android users using Android L or earlier, where the necessary APIs to do this on a per-notification basis are not available.

@skddc It might be, but my concern is that it may be confusing when this happens since it'll be displayed in a different context, much smaller and often with a mask applied.

@kenchris We shied away from this in the Notification API because you'd end up defining operating system behaviour that's out of our control. It has a note indicating that certain resources can be modified by the OS to match platform consistency.

**Comment by marcoscaceres** [comment]
@beverloo, mabye we want is to enhance image objects with a "purpose" or "role" (though 
"role" usually refers to ARIA role) and "platform" member... so:

```JSON
{
   "icons": [ {
       "src": "icon/android/badge.png",
       "purpose": "badge",
       "platform": "android"
   }, {
       "src": "icon/app.png"
   }]
}
```

**Comment by skddc** [comment]
Wouldn't the platform member be rather prone to misuse? That is, some parties/platforms would probably use other platforms' icons then, because developers can't specify (or even know) all existing and future platforms (esp. the smaller ones). Furthermore, platforms might change what style they use with new major versions.

I think this should be defined more as style or maybe purpose. A small monochrome icon like this could be used on many platforms and in many places.

Maybe it also helps to look at how Apple currently solves this for their use case (even supporting an icon color): https://developer.apple.com/library/ios/documentation/AppleApplications/Reference/SafariWebContent/pinnedTabs/pinnedTabs.html

**Comment by marcoscaceres** [comment]
Related: https://github.com/w3c/manifest/issues/361#issuecomment-201049191

**Comment by marcoscaceres** [comment]
> Wouldn't the platform member be rather prone to misuse? 

Maybe all we need is the "purpose" then.

**Comment by marcoscaceres** [comment]
Just noting that Safari's pin tabs will simply convert a site's fav icon to monochrome for pinned tabs (when rel="mask-icon" is not used): 

![screenshot 2016-08-15 16 36 47](https://cloud.githubusercontent.com/assets/870154/17957611/83622cce-6ad6-11e6-9246-304fbfda588b.png)

In the above, the "Y" is using `rel="mask-icon"` and the second, fb, is not.

**Comment by marcoscaceres** [comment]
Fixed comment above.

**Comment by marcoscaceres** [comment]
Sent proposal: 
https://rawgit.com/w3c/manifest/purpose/index.html#purpose-member
